### PR TITLE
Prioritize plugins when creating DTOs

### PR DESCRIPTION
**Pull request by abdulhaq-e**
Since SqlAlchemy is able to map `dataclasses` imperatively (classical mapping), when creating a `DTO` based on a `Mapped` SqlAlchemy dataclass, the class is recognized as a dataclass only and hence the `SqlAlchemyPlugin` is not registered.

I am not 100% how necessary is this. I had weird `pydantic` errors just because I was importing `from __future__ import annotations` in my `dataclasses` models and these errors disappeared when the DTO started using the `SqlAlchemyPlugin`.

This aligns with the behaviour in `from_model_instance` and `to_model_instance` where the first check made is whether the `DTO` has a plugin or not.

# PR Checklist

- [X] Have you followed the guidelines in `CONTRIBUTING.md`?
- [X] Have you got 100% test coverage on new code?
- [ ] Have you updated the prose documentation?
- [ ] Have you updated the reference documentation?

### CODE REVIEW

File: tests/dto_factory/test_dto_factory_plugin_registratation.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env python3
```
Comment (provinzkraut) [inline tests/dto_factory/test_dto_factory_plugin_registratation.py:1]:
```suggestion

```

File: tests/dto_factory/test_dto_factory_plugin_registratation.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    "model, exclude, field_mapping, plugins, expected_dto_plugin",
+    [
+        [Person, [], {"complex": "ultra"}, [], None],
+        [VanillaDataClassPerson, [], {"complex": "ultra"}, [], None],
+        [TypedDictPerson, [], {"complex": "ultra"}, [], None],
+        [Pet, ["age"], {"species": "kind"}, [SQLAlchemyPlugin()], SQLAlchemyPlugin],
+        [Activity, [], {}, [SQLAlchemyPlugin()], SQLAlchemyPlugin],
+    ],
+)
+def test_regiteration_of_plugin(
```
Comment (provinzkraut) [inline tests/dto_factory/test_dto_factory_plugin_registratation.py:23]:
```suggestion
def test_plugin_registration(
```

**Review by provinzkraut — CHANGES_REQUESTED** [review]
Looks okay to me, but I'm not that intimately familiar with the DTO / SQLA plugin. @Goldziher, @cofin wdyt?

**Review by Goldziher — APPROVED** [review]

File: starlite/dto.py [hunk 3]
```diff
@@ -250,23 +250,26 @@ def _get_fields_from_source(
         """Convert a `BaseModel` subclass, [`TypedDict`][typing.TypedDict], `dataclass` or any other type that has a
         plugin registered into a mapping of `str` to `ModelField`.
         """
-        plugin: Optional[PluginProtocol] = None
+        fields: Optional[Dict[str, ModelField]] = None
+        plugin = get_plugin_for_value(value=source, plugins=self.plugins)
+        if plugin:
```
[the remark this answers is not in the document]
Comment (abdulhaq-e) [reply starlite/dto.py:255]: The Walrus, a favourite of mine, been adopting it slowly. Thanks for the reminder.

**Review by provinzkraut — APPROVED** [review]

File: starlite/dto.py [hunk 4]
```diff
@@ -250,23 +250,26 @@ def _get_fields_from_source(
         """Convert a `BaseModel` subclass, [`TypedDict`][typing.TypedDict], `dataclass` or any other type that has a
         plugin registered into a mapping of `str` to `ModelField`.
         """
-        plugin: Optional[PluginProtocol] = None
+        fields: Optional[Dict[str, ModelField]] = None
+        plugin = get_plugin_for_value(value=source, plugins=self.plugins)
```
Comment (Goldziher) [inline starlite/dto.py:254]:
```suggestion
```

**Comment by Goldziher** [comment]
@all-contributors add @abdulhaq-e code

**Pull request merged**